File: 22 generate parentheses/GenerateParentheses.py
class GenerateParentheses(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = []
        cur_str = ''
        self.dfs(res, cur_str, n, 0, 0)
        return res
    
    def dfs(self, res, cur_str, total, left, right):
        if left == total and right == total:
            res.append(cur_str)
            return
        if left > right:
            self.dfs(res, cur_str + ')', total, left, right + 1)
        
        if left < total:
            self.dfs(res, cur_str + '(', total, left + 1, right)

File: 22 generate parentheses/GenerateParentheses.java

class GenerateParentheses {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        if (n <= 0) {
            return res;
        }
        genParen(res, new StringBuilder(), n, 0, 0);
        return res;
    }
    
    private void genParen(List<String> res, StringBuilder sb, int total, int left, int right) {
        if (left == total && right == total) {
            res.add(sb.toString());
            return;
        }
        if (left > right) {
            sb.append(')');
            genParen(res, sb, total, left, right + 1);
            sb.deleteCharAt(sb.length() - 1);
        }
        if (left < total) {
            sb.append('(');
            genParen(res, sb, total, left + 1, right);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}